### PR TITLE
feat: enable multiple fast-element instances in browser at once

### PR DESCRIPTION
**Pull request by EisenbergEffect**
# Pull Request

## 📖 Description

Currently, you cannot have multiple copies of `@microsoft/fast-element` running on the same page at the same time. This has to do with certain core services and their private state which break each other. For example, an observable defined by one instance of FAST cannot be observed by another because the notifier cache is not shared between the two versions.

This PR attempts to extract a set of core services that can be shared across all instances of `@microsoft/fast-element` running on the same page.

This PR also tracks versions of FAST running at once.

### 🎫 Issues

No public issues. Various public and internal instances where this has been a challenge.

## 👩‍💻 Reviewer Notes

The basic approach is that fast-element now establishes a `FAST` global which can be accessed by all running instances. The global exposes a single `get` method to retrieve shared services by id. This means that whichever fast-element version is loaded first on the page gets to define the version of the service that all instances will use.

Throughout the codebase, core services first attempt to retrieve the shared service. The API facilitates a "get or create" mechanism so that if the service is not available, the caller is able to provide an implementation.

The shared services are as follows:

* Update Queue - The DOM update queue that is used by the binding system.
* Observable - The observable system that is the core of observing state changes in templates.
* Context Event - Controls what event is available through the execution context within an event binding.
* FAST Element Definitions - The registry that maps web components to their FASTElementDefinition instances.

The API report shows breaking API changes but that is only a change from several functions on objects to properties that are functions on object. The invocation pattern doesn't change for consumers, so this isn't really a breaking change.

## 📑 Test Plan

Testing this is a challenge. All of our existing tests pass, but those don't really exercise the scenario that this is a fix for. To test this, we'd need to have two separate bundles with two versions of fast-element, on the same page at once. We don't have infrastructure setup for doing that sort of test in this repo.

To address this, I've created a manual test: [FAST Multi Instance Test.zip](https://github.com/microsoft/fast/files/8187339/FAST.Multi.Instance.Test.zip).

The attached example loads two copies of fast-element simultaneously. Two web components are defined, one with each version of the library. All code is vanilla JS, defined in script tags of type module.

#### Console Output

The console shows that the observable service used by both fast-element copies is the same instance and that there are two versions of fast-element loaded.

#### Additional Checks

* Both web components leverage attribute updates and array observation internally.
* The second web component obtains a reference to the first web component. The second component's observation mechanism is used to observe an attribute update on the first component.

### General

- [x] I have included a change request file using `$ yarn change`
- [x] I have added tests for my changes.
- [x] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/community/code-of-conduct/#our-standards) for this project.

## ⏭ Next Steps

This PR only attempts to handle scenarios where there are multiple copies of fast-element present. There may be additional issues with multiple copies of `fast-foundation`. That would need to be explored separately. Any fixes there would be dependent on the version of fast-element with this set of changes.

Furthermore, there are some things that are just not fixable, such as cross version `instanceof` checks or libraries that attempt to call `customElements.define` with the same tag name. No attempt will be made to hack around those platform restrictions.

These changes will need to be ported to FASTElement 2.0 as well. It is unclear at this time wether we will be able to support both FE1 and FE2 on the same page with this mechanism. If we can keep these core services compatible, we will. But we don't know if that's possible just yet.

As part of FE 2.0 we're hoping to put a better error/warn message system in place. Our infrastructure doesn't make that easy right now so we're delaying the warn about multiple versions until then.

### CODE REVIEW

File: packages/web-components/fast-element/src/platform.ts [hunk 1]
```diff
@@ -79,6 +104,56 @@ if ($global.trustedTypes === void 0) {
...
+                found = initialize ? (storage[id] = initialize()) : null;
+            }
+
+            return found;
+        },
+        ...propConfig,
+    });
+}
+
+// TODO: Replace with ??= after TS 4.x update
```
Comment (janechu) [inline packages/web-components/fast-element/src/platform.ts:137]: Can we put the issue https://github.com/microsoft/fast/issues/5198 in here for easier searchability?

File: packages/web-components/fast-element/src/platform.ts [hunk 2]
```diff
@@ -79,6 +104,56 @@ if ($global.trustedTypes === void 0) {
...
+($global.FAST.versions ?? (($global.FAST as any).versions = [])).push("1.8.0");
```
Comment (janechu) [inline packages/web-components/fast-element/src/platform.ts:138]: is this going to be something we need to keep updated?

Comment (EisenbergEffect) [reply packages/web-components/fast-element/src/platform.ts]: Yep. I looked at a rollup plugin to insert the version but then realized that it obviously doesn't handle the non-rollup situation. So, we might have to write a custom task that just uses regex against the file to shove the package number in there during build or something.

**Review by janechu — APPROVED** [review]

File: packages/web-components/fast-element/src/platform.ts (hunk 2 above)
Comment (nicholasrice) [reply packages/web-components/fast-element/src/platform.ts]: Are we going to do that as part of this PR or take as a follow up?

**Review by nicholasrice — APPROVED** [review]

File: packages/web-components/fast-element/src/platform.ts (hunk 2 above)
Comment (EisenbergEffect) [reply packages/web-components/fast-element/src/platform.ts]: I'm going to see if I can maybe get this done tomorrow. I'll ping here for a quick review of that part.

Comment (chrisdholt) [reply packages/web-components/fast-element/src/platform.ts]: @eisenbergeffect just seeing this - I'm going to need to look at the internals of beachball if this is still in question.

Comment (EisenbergEffect) [reply packages/web-components/fast-element/src/platform.ts]: After some investigation Chris and I discovered that the version injection won't work with beachball. I'm removing this feature for now.

**Pull request merged**